aws/resource_aws_macie2_account.go

```diff
@@ -0,0 +1,158 @@
+func resourceMacie2AccountCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*AWSClient).macie2conn
+
+	input := &macie2.EnableMacieInput{
+		ClientToken: aws.String(resource.UniqueId()),
+	}
+
+	if v, ok := d.GetOk("finding_publishing_frequency"); ok {
+		input.FindingPublishingFrequency = aws.String(v.(string))
+	}
+	if v, ok := d.GetOk("status"); ok {
+		input.Status = aws.String(v.(string))
+	}
+
+	var err error
+	err = resource.RetryContext(ctx, 4*time.Minute, func() *resource.RetryError {
+		_, err = conn.EnableMacieWithContext(ctx, input)
+		if err != nil {
+			if tfawserr.ErrCodeEquals(err, macie2.ErrorCodeClientError) {
+				return resource.RetryableError(err)
+			}
+
+			return resource.NonRetryableError(err)
+		}
+
+		return nil
+	})
+
+	if isResourceTimeoutError(err) {
+		_, err = conn.EnableMacieWithContext(ctx, input)
+	}
+
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error enabling Macie Account: %w", err))
+	}
+
+	d.SetId(meta.(*AWSClient).accountid)
+
+	return resourceMacie2AccountRead(ctx, d, meta)
+}
+
+func resourceMacie2AccountRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*AWSClient).macie2conn
+
+	input := &macie2.GetMacieSessionInput{}
+
+	resp, err := conn.GetMacieSessionWithContext(ctx, input)
+
+	if tfawserr.ErrCodeEquals(err, macie2.ErrCodeAccessDeniedException) {
+		log.Printf("[WARN] Macie not enabled for AWS account (%s), removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error reading Macie Account (%s): %w", d.Id(), err))
+	}
+
+	d.Set("status", resp.Status)
+	d.Set("finding_publishing_frequency", resp.FindingPublishingFrequency)
+	d.Set("service_role", resp.ServiceRole)
+	d.Set("created_at", aws.TimeValue(resp.CreatedAt).Format(time.RFC3339))
+	d.Set("updated_at", aws.TimeValue(resp.UpdatedAt).Format(time.RFC3339))
+
+	return nil
+}
+
+func resourceMacie2AccountUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*AWSClient).macie2conn
+
+	input := &macie2.UpdateMacieSessionInput{}
+
+	if d.HasChange("finding_publishing_frequency") {
+		input.FindingPublishingFrequency = aws.String(d.Get("finding_publishing_frequency").(string))
+	}
+
+	if d.HasChange("status") {
+		input.Status = aws.String(d.Get("status").(string))
+	}
+
+	_, err := conn.UpdateMacieSessionWithContext(ctx, input)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error updating Macie Account (%s): %w", d.Id(), err))
+	}
+
+	return resourceMacie2AccountRead(ctx, d, meta)
+}
+
+func resourceMacie2AccountDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*AWSClient).macie2conn
+
+	input := &macie2.DisableMacieInput{}
+
+	_, err := conn.DisableMacieWithContext(ctx, input)
+	if err != nil {
+		if tfawserr.ErrCodeEquals(err, macie2.ErrCodeAccessDeniedException) {
```

Comment:
In this case, we should check the message as well as the code. It's a weird code for the AWS API to return, since it isn't really an authorization problem. We want to distinguish between the case of Macie not being enabled vs the current user doesn't have permissions to access Macie.